torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,361 @@
+            If one leaves the unparametrized tensor, the shape of the parameter ``name``
+            might change
+        """
+        if not hasattr(self, "_tensor_name"):
+            raise ValueError("Parametrization has to be applied\
+                              before it can be removed")
+
+        param = get_parametrization(self._module, self._tensor_name)
+
+        if len(param) == 0:
+            raise ValueError("No parametrizations found on {}".format(self._tensor_name))
+        elif leave_parametrized and len(param) != 1:
+            raise ValueError("leave_parametrized=True only takes effect whenever"
+                             "there are exactly one parametrizations. {} parametrizations"
+                             "where found on {}".format(len(param), self._tensor_name))
+
+        # If it is the last one, we delete the created tensors
+        if len(param) == 1:
+            # Lezcano. test_pruning_id_consistency asserts that putting a constraint
+            # and removing it leaves the id of the parameter invariant
+            # This can only be achieved if the parametrization does not change the
+            # dimensions of the tensor. Is this invariance of id really necessary?
+            orig = self._module._parameters[self._tensor_name + "_orig"]
+            if leave_parametrized:
+                if self.mode == "forward":
+                    # We update the parametrization first to return it updated
+                    setattr(self._module, self._tensor_name, self.apply_param())
+                t = getattr(self._module, self._tensor_name)
+                if t.size() != orig.size():
+                    raise ValueError("We only support to leave_parametrized=True when "
+                                     "the parametrization does not change"
+                                     "the dimensions of the tensor.")
+                orig.data = t
+
+            del self._module._buffers[self._tensor_name]
+            del self._module._parameters[self._tensor_name + "_orig"]
+            self._module.register_parameter(self._tensor_name, orig)
+
+        # Remove hook
+        self._hook_handle.remove()
+
+
+def update_auto_parametrization(module, name, parametrizations):
+    # This is meant to be called from within Module.__getattr__
+    # Lezcano. Maybe move into module.py?
+
+    # Fetch tensor with the parametrization
+    _parameters = module.__dict__.get('_parameters')
+    # This should never be None, since it will just be called within
+    # __getattr__ context
+    if _parameters is None:
+        raise AttributeError("'{}' object has no attribute '{}'".format(
+            module.__name__, name + "_orig"))
+    name_orig = name + '_orig'
+    tensor = _parameters.get(name_orig)
+    if tensor is None:
+        raise AttributeError("'{}' object has no attribute '{}'".format(
+            module.__name__, name))
+    # Compute the parametrization
+    for param in parametrizations:
```

Comment:
This assumes a lot of interoperability between parametrizations. This is fine for pruning, because the pruning module handles how to exactly combine successive pruning calls, but this may not be true in the general reparametrization case.

Reply:
It is left to the user to implement these in a compatible way. It does not seem wise to impose this artificially, given that there are plenty of use cases where this could be useful. For example, in the case of wanting to put constraints on a matrix, you may choose to have a constraint that makes the matrix positive definite, and then combine this with a constraint that makes the matrix positive definite with determinant one. This makes these very flexible and easy to mix and match.